ui/src/utils/math-utils.ts

```diff
@@ -111,3 +111,12 @@ export function distToSegmentSquared(p: Point, start: Point, end: Point) {
         y: start.y + t * (end.y - start.y),
     });
 }
+
+export function linearFunctionFromPoints(
+    { x: x1, y: y1 }: Point,
+    { x: x2, y: y2 }: Point,
+    intercept: number,
```

Comment:
Tänne näytetään antavan nyt kutsupaikoissa käytännössä y1, mikä lienee virhe jos tätä nyt oikein luen. Eli eikös tässä pitäisi olla y:n arvo kohdassa x=0 (eli vaikkapa `slope*x1+intercept=y1` -> `intercept=y1-slope*x1`). Sinällään se on siis myös turha argumenttina, koska 2 pistettä määrittelee viivan ilmankin kolmatta argumenttia.

Reply:
Hmm. Lienee vähän näkökulmaeroja miten näitä tulkitsee. 

Kyseessähän on siis lineaarisen funktion yleinen muoto ( f(x) = ax + b /  f(x) = mx + b / f(x) = kx +b,  jokainen viittaa samaan asiaan). Pisteiden perusteella lasketaan suoralle kulmakerroin, joka riippumatta pisteiden arvoista kulkee funktion kuvaajassa aina origon (0, 0) kautta, kun vakiotermi jätetään huomiotta (eli intercept=0). Eli kun x=0 ja intercept=0, niin y=0, riippumatta mistä pisteistä kulmakerroin muodostettiin. Vastaavasti kun x=0, niin y = intercept. Vakioterminhän voi siis mieltää siirtävän funktion kuvaajan korkeutta pystysuunnassa (ja terminä "intercept" on mielestäni yleisimmin käytetty tällaisessa yhteydessä). 

Vakiotermin voi toki laskea tiettyä tapausta kohden, mutta silloin käyttäjä ei voi helposti määrittää "kuinka korkealla" funktion (kuvaajan) halutaan kulkevan. Pystygeometrian käyttötarkoituksessa tietenkin halutaan että f(0) = korkeus pystygeometriatiedon loppupisteessä, mutta tavoitteena tässä on määrittää yksittäisstapausta yleiskäyttöisempi funktio.

Esimerkiksi pystygeometrian kuvaajan yhteydessä varsinaiset korkeusarvot ovat alempana kuin vaikkapa se siirtoviiva, mikä käyttäjälle piirretään tilanteessa, jossa korkeusarvoja ei tunneta suunnitelmien välillä. Jos kuvaajan skaalauskäyttäytymistä haluttaisiin muuttaa huomioimaan varsinainen käyttäjälle näytettävän siirtoviivan korkeus kuvaajan reunalla (joka ei ole korkeusakselin koon yhteydessä nykytilanteessa määritetty), niin muutos voitaisiin tehdä vaihtamalla vakiotermiä. Tämä on siis yksinkertaisempaa kuin jokaisen pisteen arvon siirtäminen erikseen ylöspäin, jotta saataisiin eri kohtaan muodostuva suora, jonka avulla voitaisiin laskea reuna-arvoja.

Erot myöskään lineaarisen yhtälön ja lineaarisen funktion välillä eivät tosiaan kovin suuria ole ja tätä hyödyntävän laskun, kuten monen muunkin asian voi toteuttaa melko monella eri tavalla.

Se mikä tästä puuttuu on interceptin default-arvon asettaminen nollaksi 😄

Reply:
Aivan, kyllähän tuossa argumentissa noin on järkeä. Ajattelin asiaa käyttöpaikan suunnalta olettaen että kyseessä on pisteiden valuminen alueesta ulos ja halutaan niiden väliltä lineaarisella interpolaatiolla joku arvo. Olettaen että alkuperäiset pisteet olivat sillä korkeudella mitä haluttiin, myös niiden välistä poimitun arvon pitäisi olla, ja viivan ylös-alas siirtelylle ei pitäisi olla tarvetta.

Intercept on sinällään terminä ihan hyvä ja ymmärrettävissä mitä sillä voi tehdä, mutta se mikä tuossa itseäni hämäsi oli että miksi käyttöpaikoissa annetaan ensimmäisen pisteen Y. Tuo viivafunktiohan ei silloin mene alkuperäisten pisteiden läpi (ellei ensimmäisen pisteen X satu olemaan 0) vaan niin että se ohittaa `x=0`:n sillä korkeudella mikä tuolla ensimmäisellä pisteellä sattui olemaan (kohdassa `x=?`). Miksi?

Summa summarum, jos tuolle siirtämiselle on tarve (yleiskäyttöisyyttä kannattaa yleensä koodata vasta sitten kun sille on yleistä käyttöä) niin kolmas argumentti tietty perusteltu. En ehkä vain tajunnut tuota käyttöpaikkaa oikein.

Sinällän itse funktiosta: jos tuolle laittaa default-arvon ja käyttö näyttää siis tältä `linearFunctionFromPoints(point1, point2)`, niin ainakaan itse en olettaisi että tuon palauttama funktio läpäisee origon vaan että se läpäisee nuo molemmat pisteet. Ehkä tuo on vain koodarin vinkkeli vs matemaatikon :shrug: Sinällään ei tuossa ole mitään epäselvää jos tuo intercept on aina pakollisena mukana, koska silloin käyttäjä joutuu kiinnittämään siihen huomiota ja termi on kyllä ihan oikea.  Toinen tapa tehdä ajatusmallin ero eksplisiittiseksi olisi tietty erottaa funktiot `linearFunction(k, b)` ja `slope(p1, p2)`, jolloin tämän toiminnallisuuden saa `linearFunction(slope(p1, p2), intercept)`  ja 0 default-arvonkin kanssa (`linearFunction(slope(p1,p2))`) on edelleen selvää että viivan y-sijoittelu ei tule pisteistä. Ehkä tosin vähän tarpeetonta hienostella tuosta sen enempiä.

Reply:
Jeps. Mun aivoillahan suoran määritelmä (ns. "joukko kaikista suorista"), tietyn suoran valinta (kulmakertoimen määritys), kiinnittäminen koordinaatiston tiettyyn kohtaan (interceptin määritys eli suoran siirtäminen ylös/alas/oikealle/vasemmalle) ja se mitä arvot tietyssä yhteydessä tarkoittavat (sovelluskonteksti) ovat eri hetkiltä peräisin.

Ideanahan tuon x=0:n kiinnityksessä tällä tavoin oli pyrkiä yhdistämään x-akselin arvojen tulkinta siihen pisteeseen mistä intuitiivisesti tuo suora lähtee etenemään kohti kuvaajan laitaa. Näin muodostetaan siis ns. "lokaali koordinaatisto" siitä pisteestä x=0 lähtien mihin tuo tunnettu korkeusarvo y_1 kiinnitetään. Näin positiiviset (x-akselin) arvot voi tulkita "etäisyytenä tietystä korkeuspisteestä lähtien" ja laskemalla funktion arvo saadaan missä tahansa halutussa x-akselin pisteessä interpoloitu tai ekstrapoluitu arvo (joko negatiiviseen tai positiiviseen suuntaan). 

Näinkin muodostettu suora kulkee kyllä tunnettujen pisteiden läpi, kun lokaalin koordinaatistomuunnoksen ottaa huomioon. Esimerkiksi loppupisteen arvo laskettaisiin f(loppupiste_x - alkupiste_x), koska silloin funktion käsittelemä x:n arvo on etäisyys aloituspisteestä lähtien (eikä etäisyys vaikkapa radan aloituspisteestä). Eli kun tunnetaan pisteet $(x_1, y_1)$ ja $(x_2, y_2)$ ja käytetään tämän pullarin menetelmää, niin identtisyyksien $f(x_1 - x_1) = f(0)$ = $y_1$ sekä $f(x_2 - x_1) = y_2$ pitäisi toteutua ja tämän voisi testikeissillä varmistaakin (liukulukuvirheitä toki tapahtuu). Erityisesti siis funktioille passattua x-arvoa "siirretään" vastaamaan lokaalia koordinaatistoa (huom siis $-x_1$ termi kummassakin).

Yhtälailla nuo suorafunktiot voisi muodostaa siten, että käytettäisiin "globaalin x-akselin" (eli pystygeometrian tapauksessa ratametriosoitteita nollasta lähtien, x olisi aina >= 0) ja tällöin positiivisten arvojen tulkinta olisi tietysti "etäisyys radan alusta". Jokaisella funktiolla olisi tässä tapauksessa sama koordinaatisto. 

Se että onko lokaali vai globaali koordinaatisto intuitiivisempi riippuu taatusti ihmisestä ja kontekstista, mutta itse koin lokaalin tässä yhteydessä helpommaksi tulkita: Datasta voi hakea tietyn korkeuspisteen, katsoa kuvaajasta missä se sijaitsee, ja intuitiivisesti miettiä mihin kohtaan vaikkapa siitä lähtien piirretty suora osuu, riippumatta mistä radan kohdasta lähdettiin. Myöskään se mikä visuaalisesti kuvaajaan piirretään ei siis ole yhtenäinen (ainakaan nyt) sen suoran kanssa, jonka avulla korkeusakselin ääriarvoja määritetään (tämän pullarin aihe).

Tuo tarpeettoman geneerisyyden välttäminen on toki yleisesti tavoite. Matikkakoodia lukiessa tosin rupeaa ainakin meikäläinen välittömästi ihmettelemään, jos koodi ei noudatakaan yleisesti tunnettua matemaattista määritelmää vaikka nimityksistä saattaisikin näin päätellä. Funktiot on helpompi tunnistaa, niiden toiminta on helpompi varmistaa ja niiden käyttöperiaatteeseen voi luottaa, jos ne perustuvat matemaattiseen määritelmään, eikä niinkään käyttötarkoituksesta peräisin olevien tarpeiden perusteella muodostettuun osafunktioon, funktiomuunnokseen tai muuhun mutkansuoristukseen. Mun mielipiteenä siis matemaattinen määritelmä -> sovellus, kuin sovellus -> jokin matemaattista määritelmää melkein imitoiva funktio.

Tuo slope-funktio tosiaan auttaisi tulkintaa tässä termisekamelskassa. Lineaarinen interpolaatio on siis nimenomaan noiden pisteiden välillä toimiva menetelmä, joka on siis taas hieman eri asia kuin yleisempi lineaarinen funktio (jonka voi myös muodostaa pisteisiin perustuen, kuten tässä pullarissa tehtiin).